woocommerce/woocommerce
# Add editor history and undo/redo toolbar buttons

### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

Adds undo/redo buttons and an editor history state handler to the iframe editor.

I originally attempted to do this with the default undo/redo behavior in Gutenberg, however I ran into 2 blockers that prevented this and led to a custom solution:
1. The entity data store needs an entity to be hooked up.  Since this editor is for a piece of an entity (the description of a product), it is not tied to a specific entity.  Attempting to hook it up to an entity provider will place the entire entity template inside the iframe editor.
2. There's also the `core/editor` store which does not need to be coupled with an entity, but this selects and dispatches to the `core/data` store.  I had assumed that creating a `core/editor` subregistry would work, but since this just handles history from `core/data` it does not help unless we can also create a subregistry of that store.  That store's config does not seem to be exposed which ultimately led to being blocked from using the GB data stores for this.
<img width="411" alt="Screen Shot 2023-06-01 at 6 43 53 PM" src="https://github.com/woocommerce/woocommerce/assets/10561050/3273ef41-1358-4123-962a-41b4e466c965">



Closes #38488  .



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Enable the new product block editor under WooCommerce -> Settings -> Advanced -> Features.
2. Navigate to the Add product page
3. Open the iframe editor by clicking "Add description" under the General tab.
4. Note that the undo/redo buttons are disabled since no changes have yet been made
5. Make some changes and note the undo button is now enabled
6. Attempt to undo/redo edits made (note that the exact state of a persistent edit is determined by Gutenberg's `onChange` callback and does not always seem expected, but that is outside the scope of this PR).
7. After undoing some behavior, make some new edits and note that the redo button is no longer enabled
8. Optionally, make over 50 persistent edits (or manually turn this value down in `DEFAULT_MAX_HISTORY`) and note that you cannot undo beyond 50 edits ago
